src/ProductManifold.jl

```diff
@@ -0,0 +1,458 @@
+end
+
+Base.dataids(x::ProductArray) = Base.dataids(x.data)
+
+"""
+    find_pv(x...)
+
+`A = find_pv(x...)` returns the first `ProductArray` among the arguments.
+"""
+@inline find_pv(bc::Base.Broadcast.Broadcasted) = find_pv(bc.args)
+@inline find_pv(args::Tuple) = find_pv(find_pv(args[1]), Base.tail(args))
+@inline find_pv(x) = x
+@inline find_pv(a::ProductArray, rest) = a
+@inline find_pv(::Any, rest) = find_pv(rest)
+
+size(x::ProductArray) = size(x.data)
+Base.@propagate_inbounds getindex(x::ProductArray, i) = getindex(x.data, i)
+Base.@propagate_inbounds setindex!(x::ProductArray, val, i) = setindex!(x.data, val, i)
+
+(+)(v1::ProductArray{ShapeSpec}, v2::ProductArray{ShapeSpec}) where ShapeSpec<:ShapeSpecification = ProductArray(ShapeSpec, v1.data + v2.data)
+(-)(v1::ProductArray{ShapeSpec}, v2::ProductArray{ShapeSpec}) where ShapeSpec<:ShapeSpecification = ProductArray(ShapeSpec, v1.data - v2.data)
+(-)(v::ProductArray{ShapeSpec}) where ShapeSpec<:ShapeSpecification = ProductArray(ShapeSpec, -v.data)
+(*)(a::Number, v::ProductArray{ShapeSpec}) where ShapeSpec<:ShapeSpecification = ProductArray(ShapeSpec, a*v.data)
+
+eltype(::Type{ProductArray{TM, TData, TV}}) where {TM, TData, TV} = eltype(TData)
+similar(x::ProductArray{ShapeSpec}) where ShapeSpec<:ShapeSpecification = ProductArray(ShapeSpec, similar(x.data))
+similar(x::ProductArray{ShapeSpec}, ::Type{T}) where {ShapeSpec<:ShapeSpecification, T} = ProductArray(ShapeSpec, similar(x.data, T))
+
+"""
+    ProductMPoint(parts)
+
+A more general but slower representation of points on a product manifold.
+"""
+struct ProductMPoint{TM<:Tuple} <: MPoint
+    parts::TM
+end
+
+ProductMPoint(points...) = ProductMPoint{typeof(points)}(points)
+eltype(x::ProductMPoint) = eltype(Tuple{map(eltype, x.parts)...})
+similar(x::ProductMPoint) = ProductMPoint(map(similar, x.parts)...)
+similar(x::ProductMPoint, ::Type{T}) where T = ProductMPoint(map(t -> similar(t, T), x.parts)...)
+
+function proj_product(x::ProductMPoint, i::Integer)
+    return x.parts[i]
+end
+
+"""
+    ProductTVector(parts)
+
+A more general but slower representation of tangent vectors on a product
+manifold.
+"""
+struct ProductTVector{TM<:Tuple} <: TVector
+    parts::TM
+end
+
+ProductTVector(vectors...) = ProductTVector{typeof(vectors)}(vectors)
+eltype(x::ProductTVector) = eltype(Tuple{map(eltype, x.parts)...})
+similar(x::ProductTVector) = ProductTVector(map(similar, x.parts)...)
+similar(x::ProductTVector, ::Type{T}) where T = ProductTVector(map(t -> similar(t, T), x.parts)...)
+
+(+)(v1::ProductTVector, v2::ProductTVector) = ProductTVector(map(+, v1.parts, v2.parts)...)
+(-)(v1::ProductTVector, v2::ProductTVector) = ProductTVector(map(-, v1.parts, v2.parts)...)
+(-)(v::ProductTVector) = ProductTVector(map(-, v.parts))
+(*)(a::Number, v::ProductTVector) = ProductTVector(map(t -> a*t, v.parts))
+
+function proj_product(x::ProductTVector, i::Integer)
+    return x.parts[i]
+end
```

Comment:
similar as above – and why do you then _not_ copy here?

Reply:
Let's discuss it above and I'll change both methods to make them consistent.